File: glass.R
library(ISLR)
glass=read.csv("E:/Assignments/ASsignment week 12/KNN/Assignment/glass.csv")
str(glass)
glass$Type=as.factor(glass$Type)
summary(glass$Type)
purchase=glass[,10]
stglass=scale(glass[,-10])
# First 100 rows for test set
test.index <- 1:40
test.data <- stglass[test.index,]
test.purchase <- purchase[test.index]
# Rest of data for training
train.data <- stglass[-test.index,]
train.purchase <- purchase[-test.index]
library(class)
set.seed(101)
predicted.purchase <- knn(train.data,test.data,train.purchase,k=10)
head(predicted.purchase)
mean(test.purchase != predicted.purchase)
predicted.purchase <- knn(train.data,test.data,train.purchase,k=3)
mean(test.purchase != predicted.purchase)
predicted.purchase <- knn(train.data,test.data,train.purchase,k=5)
mean(test.purchase != predicted.purchase)

File: glass.py
import pandas as pd
import numpy as np

glass=pd.read_csv("E:/Assignments/ASsignment week 12/KNN/Assignment/glass.csv")

def norm(i):
    x=((i-i.min())/ (i.max()-i.min()))
    return x
glass_norm=norm(glass.iloc[:,0:9])

glass_norm.describe()

X = np.array(glass_norm.iloc[:,:]) # Predictors 
Y = np.array(glass['Type']) 
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 10)
knn.fit(X_train, Y_train)
pred = knn.predict(X_test)
pred
# Evaluate the model
from sklearn.metrics import accuracy_score
print(accuracy_score(Y_test, pred))
pd.crosstab(Y_test, pred, rownames = ['Actual'], colnames= ['Predictions'])
# error on train data
pred_train = knn.predict(X_train)
print(accuracy_score(Y_train, pred_train))
pd.crosstab(Y_train, pred_train, rownames=['Actual'], colnames = ['Predictions']) 
# creating empty list variable 
acc = []
for i in range(3,50,2):
    neigh = KNeighborsClassifier(n_neighbors=i)
    neigh.fit(X_train, Y_train)
    train_acc = np.mean(neigh.predict(X_train) == Y_train)
    test_acc = np.mean(neigh.predict(X_test) == Y_test)
    acc.append([train_acc, test_acc])
    
    
import matplotlib.pyplot as plt # library to do visualizations 

# train accuracy plot 
plt.plot(np.arange(3,50,2),[i[0] for i in acc],"ro-")

# test accuracy plot
plt.plot(np.arange(3,50,2),[i[1] for i in acc],"bo-")